Sources/Foundation/Morphology.swift

```diff
@@ -0,0 +1,349 @@
+            return self[keyPath: path]
+        } else {
+            return nil
+        }
+    }
+}
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+extension Morphology {
+    public func customPronoun(forLanguage language: String) -> CustomPronoun? {
+        return customPronouns[language.lowercased()]
+    }
+    
+    mutating public func setCustomPronoun(_ pronoun: CustomPronoun?, forLanguage language: String) throws {
+        if let pronoun = pronoun {
+            let result = pronoun.validate(forLanguage: language)
+            if !result {
+                throw CocoaError(.keyValueValidation)
+            }
+        }
+        
+        customPronouns[language.lowercased()] = pronoun
+    }
+
+    @available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+    public struct CustomPronoun {
+        public init() {}
+
+        public static func isSupported(forLanguage language: String) -> Bool {
+            let indexAtTwo = language.index(language.startIndex, offsetBy: 2)
+            // For now we only support English custom pronoun
+            return language.count >= 2 &&
+                language.lowercased().starts(with: "en") &&
+                (language.count == 2 ||
+                 language[indexAtTwo ..< language.index(after: indexAtTwo)] == "-" ||
+                 language[indexAtTwo ..< language.index(after: indexAtTwo)] == "_")
+        }
+
+        public static func requiredKeys(forLanguage language: String) -> [PartialKeyPath<Self>] {
+            guard self.isSupported(forLanguage: language) else {
+                return []
+            }
+            return [\.subjectForm, \.objectForm, \.possessiveForm, \.possessiveAdjectiveForm, \.reflexiveForm]
+        }
+        
+        fileprivate func validate(forLanguage language: String) -> Bool {
+            guard Self.isSupported(forLanguage: language) else {
+                return false
+            }
+            
+            for keyPath in Self.requiredKeys(forLanguage: language) {
+                let value: Any? = self[keyPath: keyPath]
+                if value == nil {
+                    return false
+                }
+            }
+            
+            return true
+        }
+
+        public var subjectForm: String?
+        public var objectForm: String?
+        public var possessiveForm: String?
+        public var possessiveAdjectiveForm: String?
+        public var reflexiveForm: String?
+    }
+}
+
+// MARK: -
+// MARK: NSAttributedString Keys
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+extension NSAttributedString.Key {
+    internal static var morphology = NSAttributedString.Key(rawValue: "NSMorphology")
+    internal static var inflectionRule = NSAttributedString.Key(rawValue: "NSInflect")
+    internal static var inflectionAlternative = NSAttributedString.Key(rawValue: "NSInflectionAlternative")
+    internal static var language = NSAttributedString.Key(rawValue: "NSLanguage")
```

Comment:
Also, they should be `let's.

```suggestion
    public static let morphology = NSAttributedString.Key(rawValue: "NSMorphology")
    public static let inflectionRule = NSAttributedString.Key(rawValue: "NSInflect")
    public static let inflectionAlternative = NSAttributedString.Key(rawValue: "NSInflectionAlternative")
    public static let language = NSAttributedString.Key(rawValue: "NSLanguage")
```